walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,202 @@
+      return true;
+    }
+    return false;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<TimeRange> availableTimeRanges = new ArrayList<TimeRange>();
+
+    if (request.getDuration() > (long)TimeRange.WHOLE_DAY.duration()){
+      return availableTimeRanges;
+    }
+
+    if (events.isEmpty()){
+      availableTimeRanges.add(TimeRange.WHOLE_DAY);
+      return availableTimeRanges;
+    }
+
+    HashMap<TimeRange, Event> timeRangeEventMap = new HashMap<TimeRange, Event>();
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>();
+
+    for (Event event : events){
+      timeRangeEventMap.put(event.getWhen(), event);
+      timeRanges.add(event.getWhen());
+    }
+    timeRanges.sort(TimeRange.ORDER_BY_START);
+
+    boolean onlyOptionalEvents = false;
+
+    // Remove Optional Events from iterator (timeRanges)
+    ArrayList<Event> optionalEvents = new ArrayList<Event>();
+    ArrayList<TimeRange> removeOpRanges = new ArrayList<TimeRange>();
+
+    for(TimeRange timeRange : timeRanges){
+
+      Event currentEvent = timeRangeEventMap.get(timeRange);
+
+      if (optionalAttendeesEvent(request, currentEvent)){
+        optionalEvents.add(currentEvent);
+        removeOpRanges.add(timeRange);
+      }
+    }
+
+    for (TimeRange opTimeRange : removeOpRanges){
+      timeRanges.remove(opTimeRange);
+    }
+
+
+    if (timeRanges.isEmpty()){
+      onlyOptionalEvents = true;
+      timeRanges = removeOpRanges;
+    }
+
+
+    int avaliableStart = TimeRange.START_OF_DAY;
+    int avaliableEnd;
+
+    for (int index = 0; index < timeRanges.size(); index++){
+
+      Event currentEvent = timeRangeEventMap.get(timeRanges.get(index));
+
+      if (eventNoMandatoryAttendees(request, currentEvent)){
+        if (timeRanges.size() == 1){
+          avaliableEnd = TimeRange.END_OF_DAY;
+          availableTimeRanges.add(TimeRange.fromStartEnd(avaliableStart, avaliableEnd, true));
+        }
+        if (!onlyOptionalEvents){
+          continue;
+        }
+      }
+
+      avaliableEnd = timeRanges.get(index).start();
+      
+      // [##, ##)
+      TimeRange avaliableTimeExclusive = TimeRange.fromStartEnd(avaliableStart, avaliableEnd, false);
+      // [##, ## + 1)
+      TimeRange avaliableTimeInclusive = TimeRange.fromStartEnd(avaliableStart, avaliableEnd, true);
+
+      // Valid TimeRanges Check
+      if (!(avaliableTimeExclusive.overlaps(timeRanges.get(index)) || avaliableTimeExclusive.duration() < request.getDuration())){
+        availableTimeRanges.add(avaliableTimeExclusive);
+      }
+
+      if(!(avaliableTimeInclusive.overlaps(timeRanges.get(index)) || avaliableTimeInclusive.duration() < request.getDuration())){
+        availableTimeRanges.add(avaliableTimeInclusive);
+      }
+
+      // Nested Events Check
+      if (index != 0){
```

Comment:
```//
    // Events  :       |----A----|
    //                   |--B--|
    // Day     : |---------------------|
    // Options : |--1--|         |--2--|
```

Hi Vivek, the ```Nested Events Check```  needs to create a new ```TimeRange``` because else it would output an incorrect Collection of ```TimeRange```

In this example, if I didn't have the check it would output:

```[Range: [0, 510), Range: [570, 1440)]```

The second range is incorrect. Without this check block, the algorithm start the range at the
end of ```Event B```, yet it should start at the end of ```Event A```.

The correct and expected output is: 

```[0, 510), Range: [600, 1440)]```

Reply:
Thanks for the explanation Mauro! 

Would your algorithm still work even on the following example? 
```//
    // Events  :       |----A----|
    //                   |--B--|
    //                    |-C-|
    // Day     : |---------------------|
    // Options : |--1--|         |--2--|
```

Reply:
I don't think my algorithm would work for this scenario. Thus, I created an additional check if  ```availableStart > availableEnd```. I also created a test for this scenario, which passes after making the previously mention check. Let me know if I made an mistake, I think it should work for all nested events

Reply:
Nice! Yeah, adding the check for `availableStart > availableEnd` was exactly what I had in mind as well :)